---
title: "Clarava Visualizations: DeSeq2 Differential Gene Expression"
author: "VericiDx Bioinformatics"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE)
knitr::opts_chunk$set(cache= TRUE)
```

This R Notebook calculates differential expressed genes (DEG) using DESeq2 and generates volcano plot for Clarava Manuscript.

# Install and load required packages

```{r}
library(tidyverse)
library(DESeq2)
library(biomaRt)
library(yaml)
library(pheatmap)
library(Glimma)
library(highcharter)
library(EnhancedVolcano)
library(clusterProfiler)
```

# Load the data
Read config file and define file paths for inputs. The inputs includes the gene counts, annotation and gene lists. Finally, we read the inputs.
```{r define paths}
# prepare config file and directories
config_path <- "../config.yml"

config <- read_yaml(config_path)

#Base directory
base_dir <- dirname(normalizePath(config_path))

#Constants directory - this contains subdirectories including the gene lists and counts
data_dir <- file.path(base_dir, config$data_dir)

#Other directories with constant files
gene_lists_dir <- file.path(paste(data_dir, config$gene_lists_dir, sep = ""))
counts_dir <- file.path(paste(data_dir, config$counts_dir, sep = ""))
annotations_dir <- file.path(paste(data_dir, config$annotations_dir, sep = ""))
results_dir <- file.path(paste0(base_dir, "/",config$results_dir))

#Annotations for the training samples
annotation_data_path <- file.path(paste(annotations_dir, config$ct1_annotations_123samples, sep = ""))

# Raw gene data after the raw filter
# Following count filtering 11170 genes remained for analysis
raw_gene_data_post_rawfilter_path <- file.path(paste(counts_dir, config$gene_data_raw_counts, sep = ""))

# Gene Signature Path
gene_sig_29_path <- file.path(paste0(gene_lists_dir, config$gene_sig_29))

# Define the output path
res_out_path <- file.path(paste(results_dir, config$deseq2_results_firstRawFilter, sep = ""))
```

```{r read data}
annotation_data <- read.csv(annotation_data_path, row.names = 1)
rownames(annotation_data) <- annotation_data$Sample_ID


gene_data_raw <- read.csv(raw_gene_data_post_rawfilter_path, check.names = F, row.names = 1)
```


```{r}
#Convert to matrix and integer format
gene_matrix <- as.matrix(gene_data_raw)
mode(gene_matrix) <- "integer"
```

We limit samples to the ones in the annotation file.
There are total 123 samples and 11170 genes.
```{r}
gene_matrix <- gene_matrix[,annotation_data$Sample_ID]
dim(gene_matrix)


# Check for NA values in the matrix
any(is.na(gene_matrix))
```

Check if annotation and gene data are in the same order
```{r}
all(rownames(annotation_data) == colnames(gene_matrix))
```


```{r}
# Convert the rejection annotations to a factor
annotation_data$MM_Rejection_Determination <- as.factor(annotation_data$MM_Rejection_Determination)

# Print a table of the rejection statuses
table(annotation_data$MM_Rejection_Determination) # Checks out
```

# DESeq2 Analysis
Create a DESeqDataSet object from the count and annotation data.

```{r}
ddsMat <- DESeqDataSetFromMatrix(countData=gene_matrix, colData=annotation_data, design= ~ MM_Rejection_Determination)
```

We now perform differential expression analysis based on the negative binomial (A.K.A. Gamma-Poisson) distribution. This is done through estimation of size factors (using the "median ratio method"), estimation of dispersion and testing for significance of coefficients in a Negative Binomial GLM, using previously calculated sizeFactors (or normalizationFactors) and dispersion estimates.

Perform differential expression analysis.
```{r}
dds <- DESeq(ddsMat)
```

## Extract the results
```{r}
result <- results(dds, c("MM_Rejection_Determination", "Reject", "NoReject"))
summary(result)

results(dds)
```

We check how many genes are considered significant.

```{r}
## Extract a dataframe of the DeSeq2 results for all samples
result_object <- na.omit(result)
filtered_result <- data.frame(result_object@listData)
rownames(filtered_result) <- result_object@rownames
filtered_result[,"GeneID"] <- rownames(filtered_result)
```

Check how many values pass certain thresholds (user testing)
```{r}
# Almac cutoffs 
# a relatively high p-value cut-off of 0.01 and no fold change threshold to ensure any gene that may contribute even a small amount to discriminatory power was included.
sum(filtered_result$pvalue < 0.01) # 344
sum(filtered_result$pvalue < 0.01 & (abs(filtered_result$log2FoldChange) >= 1)) # 5
```

```{r echo = FALSE}
# Not a lot of genes with strict cutoffs
sum(filtered_result$padj < 0.1)
sum(filtered_result$padj < 0.05)
sum(filtered_result$padj < 0.01)

sum(filtered_result$padj < 0.1 & (filtered_result$log2FoldChange <= -0.5 | filtered_result$log2FoldChange >=0.5))
sum(filtered_result$padj < 0.05 & (filtered_result$log2FoldChange <= -0.5 | filtered_result$log2FoldChange >=0.5))
sum(filtered_result$padj < 0.01 & (filtered_result$log2FoldChange <= -0.5 | filtered_result$log2FoldChange >=0.5))

sum(filtered_result$padj < 0.1 & (filtered_result$log2FoldChange <= -0.75 | filtered_result$log2FoldChange >=0.75))
sum(filtered_result$padj < 0.05 & (filtered_result$log2FoldChange <= -0.75 | filtered_result$log2FoldChange >=0.75))
sum(filtered_result$padj < 0.01 & (filtered_result$log2FoldChange <= -0.75 | filtered_result$log2FoldChange >=0.75))
```

```{r}
write.csv(filtered_result, res_out_path)
```

```{r glmdplot, eval = FALSE, echo = FALSE}
#
#### User intervention - annotation data column must match the rejection column
glMDPlot(result, anno=filtered_result, samples=annotation_data$Sample_ID, groups=annotation_data$MM_Rejection_Determination, counts=gene_matrix, launch=TRUE)
```

```{r volcano plots}
# Define custom cols
# create custom key-value pairs for 'high', 'low' expression by fold-change
keyvals <- ifelse(
    filtered_result$log2FoldChange < 0 & filtered_result$pvalue<= 0.01, 'steelblue3',
      ifelse(filtered_result$log2FoldChange > 0 & filtered_result$pvalue<= 0.01, 'orange3',
        'grey'))
keyvals[is.na(keyvals)] <- 'grey'
names(keyvals)[keyvals == 'orange3'] <- 'high'
names(keyvals)[keyvals == 'steelblue3'] <- 'low'

#Volcano plot for the results
p1 <- EnhancedVolcano(filtered_result,
                      lab = rownames(filtered_result),
                      x = 'log2FoldChange',
                      y = 'pvalue',
                      pCutoff = 1e-2,
                      FCcutoff = 0,
                      pointSize = 1.0,
                      selectLab = c('TMEM176B','ADH1A'),
                      labSize = 2.0,
                      labCol = 'black',
                      boxedLabels = FALSE,
                      #col = c('black', 'grey','steelblue4', 'orange3'),
                      colCustom = keyvals,
                      title = "",
                      subtitle = "",
                      cutoffLineType= "blank", # remove cutoff lines to set manually for hline
                      hline = 1e-2,
                      # drawConnectors = TRUE,
                      # widthConnectors = 1.0,
                      # colConnectors = "black",
                      lengthConnectors = unit(0.05, 'npc'),
                      max.overlaps = Inf,
                      legendPosition = "right",
                      hlineWidth = 0.5,
                      arrowheads = FALSE,
                      ylim = c(0, 5.5),
                      xlim = c(-1.5, 1.5), 
                      gridlines.major = FALSE,
                      gridlines.minor = FALSE)
p1     

```

```{r save plot}
vp_path <- file.path(paste0(results_dir, "volcano_plot.pdf"))
pdf(vp_path, height = 10, width = 10)
p1
dev.off()

```



```{r volcano plot label 29 genes, echo = FALSE}

# Add gene symbol to filtered_result
ens <- filtered_result$GeneID
mart <- useDataset("hsapiens_gene_ensembl", mart = useMart("ensembl"))
g_list <- getBM(attributes = c("ensembl_gene_id", "hgnc_symbol"), values = ens, mart = mart, filters = "ensembl_gene_id")

filtered_result <- left_join(filtered_result, g_list, by = c("GeneID" = "ensembl_gene_id"))

# 29 gene signature
gene_sig_29 <- read.csv(gene_sig_29_path)

# Filter significant genes with p-value < 0.01 for labeling
lab_df <- filtered_result[which(filtered_result$hgnc_symbol %in% gene_sig_29$Gene.Symbol), ]

lab_df <- lab_df %>% filter(pvalue <= 0.01 )
# Volcano plot - label 29 genes signature
p2 <- EnhancedVolcano(filtered_result,
                      lab = filtered_result$hgnc_symbol,
                      x = 'log2FoldChange',
                      y = 'pvalue',
                      pCutoff = 1e-2,
                      FCcutoff = 0,
                      selectLab = gene_sig_29$Gene.Symbol,
                      pointSize = 1.0,
                      #selectLab = c('TMEM176B','ADH1A'),
                      labSize = 2.0,
                      labCol = 'black',
                      boxedLabels = FALSE,
                      #col = c('black', 'grey','steelblue4', 'orange3'),
                      colCustom = keyvals,
                      title = "",
                      subtitle = "",
                      cutoffLineType= "blank", # remove cutoff lines to set manually for hline
                      hline = 1e-2,
                      drawConnectors = TRUE,
                      # widthConnectors = 1.0,
                      # colConnectors = "black",
                      lengthConnectors = unit(0.05, 'npc'),
                      max.overlaps = Inf,
                      legendPosition = "right",
                      hlineWidth = 0.5,
                      arrowheads = FALSE,
                      ylim = c(0, 7.5),
                      xlim = c(-2, 2), 
                      gridlines.major = FALSE,
                      gridlines.minor = FALSE)
p2

```

